### PR TITLE
Fix bug in `Heat.indexing.nonzero`

### LINKED ISSUE #2136: [Bug]: Bug in `heat.indexing.nonzero`
### What happened?

The function `heat.indexing.non_zero` can return strange data in some cases. In the example below, it returns a DNDarray of shape (2,), with an larray of shape (1,2). This would correspond to kind of reverse distribution where the total global data is smaller than local data and is totally wrong.

It seems to me that this bug is causing the bug documented in #2135, so fixing this one could hopefully also fix that one.

### Code snippet triggering the error

```shell
import heat as ht

a = ht.zeros((4, 3), dtype=ht.bool)
a[1, 2] = True
non_zero = ht.indexing.nonzero(a)
print(non_zero.__repr__())
"""
DNDarray(MPI-rank: 0, Shape: (2,), Split: None, Local Shape: (1, 2), Device: cpu:0, Dtype: int64, Data:
         [[1, 2]])
"""
```

### Error message or erroneous outcome

```shell

```

### Version

main (development branch)

### Python version

3.13

### PyTorch version

2.10

### MPI version

```shell

```

### PR DESCRIPTION
**Pull request by brownbaerchen**
## Due Diligence

- General:
    - [x]  **title** of the PR is suitable to appear in the [Release Notes](https://github.com/helmholtz-analytics/heat/releases/latest)
- Implementation:
    - [x] unit tests: all split configurations tested
    - [x] unit tests: multiple dtypes tested
    - [ ] **NEW** unit tests: MPS tested (1 MPI process, 1 GPU)
    - [ ] benchmarks: created for new functionality
    - [ ] benchmarks: performance improved or maintained
    - [ ] documentation updated where needed

## Description
This fixes the bugs documented in #2136 and #2135. It makes the function `heat.indexing.nonzero` compatible with the sanity checks proposed in #2137.

Issue/s resolved: #2136, #2135 

## Changes proposed:

- Refactor and bugfix in `heat.indexing.nonzero`

## Type of change

 - bug fix

#### Does this change modify the behaviour of other functions? If so, which?
no

### CODE REVIEW

File: heat/core/tests/test_indexing.py [hunk 1]
```diff
@@ -23,6 +23,15 @@ def test_nonzero(self):
         a[nz] = 10.0
         self.assertEqual(ht.all(a[nz] == 10), 1)
 
+        # edge case: single non-zero element
+        for split in [None, 1]:
+            a = ht.zeros((4, 3), dtype=ht.bool, split=split)
+            a[1, 2] = True
+            nz = ht.indexing.nonzero(a)
+            self.assertEqual(nz.gshape, (1, 2))
+            self.assertTrue(ht.allclose(a[nz], a[a]))
```
Comment (mtar) [inline heat/core/tests/test_indexing.py]: The edge case for split=1 doesn't work. All local tensors have no data after indexing.
```
print(a.comm.rank, a[nz].larray)
print(a.comm.rank, a[a].larray)
```
```
0 tensor([], dtype=torch.bool)
1 tensor([], dtype=torch.bool)

0 tensor([], dtype=torch.bool)
1 tensor([], dtype=torch.bool)
```

**Review by mtar — CHANGES_REQUESTED** [review]
Thanks, the PR looks fine. The added edge case doesn't work properly. Something's wrong with the indexing.

File: heat/core/tests/test_indexing.py (hunk 1 above)
Comment (brownbaerchen) [reply heat/core/tests/test_indexing.py]: That's true, good catch!

This PR is kind of working on two issues. The first is in `heat.indexing.nonzero` and the second is in the advanced indexing. Turns out this fixes the advanced indexing bug only when the array is non-distributed. 
However, the primary function of this PR, which is fixing the bug in nonzero is still valid. I modified the test to index only non-distributed data. Note that this test fails on the current main.

I suggest we merge this anyways and hope that #938 will fix the indexing.

**Review by mtar — APPROVED** [review]
